spices/SPICE-0004-HTTP-Proxy-Support.adoc

```diff
@@ -0,0 +1,180 @@
+* Status: TBD
+* Implemented in: TBD
+* Category: Language, Standard Library
+
+== Introduction
+
+This SPICE describes support for HTTP proxying.
+
+== Motivation
+
+Some environments do not allow direct connections to the internet, instead requiring that connections go through an intermediary proxy.
+
+To enable Pkl in these situations, this SPICE provides a way to redirect outbound requests through said proxies.
+
+== Proposed Solution
+
+Pkl will support HTTP proxies.
+Proxy settings will be added, and this will control how Pkl makes HTTP requests.
+
+== Detailed design
+
+=== Configuring proxies
+
+A proxy can be set via the following, in order of precedence (least to greatest):
+
+1. `~/.pkl/settings` file
+2. `PklProject` file (in `evaluatorSettings`)
+3. CLI Flags/pkl-gradle options/`HttpClient#Builder`/server mode message passing API
+
+==== Environment Variables
+
+Proxies will _not_ be configured via environment variables.
+This maintains consistency with the rest of Pkl; it does not support using environment variables for any other configuration.
+Additionally, while other languages might use environment variables, they have slightly different meaning from one language to another (see link:https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy[]]).
+Not using environment variables avoids a possble problem where the same environment variable influences both Pkl and another runtime, but in different ways.
+
+==== Settings file
+
+A new `proxy` property is added to the global settings file.
+
+[source,pkl]
+----
+module pkl.settings
+
+import "pkl:Proxy"
+
+/// HTTP proxy settings.
+@Since { version = "0.26.0" }
+proxy: Proxy?
+----
+
+==== `PklProject`
+
+A new `proxy` property is added to `EvaluatorSettings` of a PklProject file.
+
+[source,pkl]
+----
+module pkl.Project
+
+import "pkl:Proxy"
+
+class EvaluatorSettings {
+  proxy: Proxy?
+}
+----
+
+[[new-stdlib-module]]
+==== New stdlib module: `pkl.Proxy`
+
+A new stdlib module is added called `pkl.Proxy`, that is used by `pkl.settings` and `pkl.Project`.
+
+[source,pkl]
+----
+/// Settings that control how Pkl talks to HTTP proxies.
+@Since { version = "0.26.0" }
+module pkl.Proxy
+
+/// The proxy to use for HTTP(s) connections.
+///
+/// At the moment, only HTTP proxies are supported.
+///
+/// Example:
+/// ```
+/// local proxyPassword = read("my-proxy-password.txt").text
+/// address = "http://myuser@\(proxyPassword):my.proxy.example.com:5080"
+/// ```
+address: Uri(startsWith("http://"))
+
+/// Hosts to disable connecting to via a proxy.
+///
+/// Values can be either hostnames, or CIDR blocks.
+///
+/// Follows the same rules as cURL's [--noproxy](https://curl.se/docs/manpage.html#--noproxy)
+/// flag.
+noProxy: Listing<String>(isDistinct)
+----
+
+==== CLI flags
+
+The following CLI flags are added to common CLI options:
+
+* `--proxy` -- A proxy URL to connect to
+* `--noproxy` -- A comma-separated list of hosts to not proxy.
+
+==== pkl-gradle
```

Comment:
What’s the use case case for having different proxy settings for Pkl than for Gradle?

Reply:
I want to control `pkl eval`  through Gradle, but I want it to otherwise behave exactly like my CLI does.

Reply:
The correct proxy settings depend on the network you’re on. I can’t think of any use case where it is useful to have different proxy settings for Gradle and the Gradle Pkl plugin.